File: app/models/database/compliance.py
from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import Boolean, String, Text, TIMESTAMP, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func

from app.models.database.base import Base


class ComplianceEvent(Base):
    __tablename__ = "compliance_events"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    organization_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("organizations.id", ondelete="CASCADE"), nullable=False, index=True
    )
    borrower_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("borrowers.id"), nullable=False)
    call_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("calls.id"))
    event_type: Mapped[str] = mapped_column(String(80), nullable=False)
    severity: Mapped[str] = mapped_column(String(20), default="info")   # info | warning | violation
    description: Mapped[str | None] = mapped_column(Text())
    auto_actioned: Mapped[bool] = mapped_column(Boolean(), default=False)
    reviewed_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))
    reviewed_at: Mapped[datetime | None] = mapped_column(TIMESTAMP(timezone=True))
    created_at: Mapped[datetime] = mapped_column(TIMESTAMP(timezone=True), server_default=func.now())

    borrower: Mapped["Borrower"] = relationship("Borrower", back_populates="compliance_events")
    call: Mapped["Call | None"] = relationship("Call", back_populates="compliance_events")

File: frontend/src/api/compliance.ts
import client from "./client";

export interface ComplianceEvent {
  id: string;
  borrower_id: string;
  call_id?: string;
  event_type: string;
  severity: "info" | "warning" | "violation";
  description?: string;
  auto_actioned?: boolean;
  created_at: string;
}

export const complianceApi = {
  getEvents: async (borrower_id: string) => {
    const res = await client.get<ComplianceEvent[]>(`/compliance/events/${borrower_id}`);
    return res.data;
  },
  listRecent: async (severity?: string) => {
    const params = severity ? { severity } : {};
    const res = await client.get<ComplianceEvent[]>("/compliance/events", { params });
    return res.data;
  },
  recordConsent: async (borrower_id: string) => {
    const res = await client.post(`/compliance/consent/${borrower_id}`);
    return res.data;
  },
};
